### ISSUE TITLE
xmap does not throw error when parameters are sharded across model axis

### ISSUE THREAD

**Issue by wilson1yan** [google/jax#13229]
### Description

I have a simple script below that uses `xmap` to initialize a set of weights, and return them. 

I would expect that it should be `out_axes=('model', ...)` since the input `rng` (which is used to init the weights) is also sharded across the `model` axis, but the code also seems to run smoothly if I set `out_axes=(...,)` and it seems to just return the corresponding weight at index 0 (along the `model` axis) even though the weight is different along each model shard.

Usually if I set `out_axes=(...,)` incorrectly when the output has not been reduced along an axis (e.g. through `pmean`, `psum`, etc), it throws an error - but it doesn't seem to in this case. Is this on purpose or is this a bug?

This was tested on a `v3-8` with `tpu-vm-base`

```python
import numpy as np
import flax.linen as nn
import jax
import jax.numpy as jnp
from jax.experimental.maps import xmap, Mesh
import flax


class Model(nn.Module):
    @nn.compact
    def __call__(self, x):
        w = self.param('weight', nn.initializers.normal(), [x.shape[-1], 2], jnp.float32)
        x = x @ w
        return jnp.mean(x ** 2)

num_shards = 2
devices = np.array(jax.devices()).reshape(jax.device_count() // num_shards, num_shards)
mesh = Mesh(devices, ['dp', 'mp'])

x = np.random.randn(1, 1)
rngs = jax.random.split(jax.random.PRNGKey(0), num_shards)
model = Model()

def init(rng, x):
    variables = model.init(
        {'params': rng},
        x,
    ).unfreeze()
    params = variables.pop('params')
    return params, params

p_init = xmap(
    init,
    in_axes=(('model', ...), (...,)),
    out_axes=(('model', ...), (...,)),
    axis_resources={'model': 'mp'}
)

with mesh:
    params_sharded, params = p_init(rngs, x)
print('return sharded:', jax.device_get(params_sharded['weight']))
print('return not sharded:', jax.device_get(params['weight']))
```

Example output:
```
return sharded: [[[-0.01583592 -0.00883907]]

 [[-0.00255863  0.01448902]]

 [[-0.00728441  0.00094846]]

 [[ 0.00259233  0.00254223]]]
return not sharded: [[-0.01583592 -0.00883907]]
```
Weights are different along each model index, but the not sharded only selects the 0th index one.

### What jax/jaxlib version are you using?

`jax==0.3.24`, `jaxlib==0.3.24`, `libtpu-nightly==0.1.dev20221103`, `flax==0.6.1`

### Which accelerator(s) are you using?

TPU

**Comment by wilson1yan** [comment]
is there any update on this?

**Comment by apaszke** [comment]
xmap was removed from JAX, so this is obsolete
